Implement --controller<x> options in ed64romconfig to change the respective Advanced Homebrew Header fields

risingPhil opened the pull request:
Implement --controller<x> options in ed64romconfig to change the respective Advanced Homebrew Header fields

Modifying these fields may influence emulator behaviour, such as Ares'

The context in which I implemented this is to make debugging N64 Transfer Pak code easier with Ares Emulator: 
After using the "--controller1 n64,pak=transfer" option on your rom, Ares will ask for a gameboy rom for the emulated transfer pak everytime you load the rom. 

(Note: In Ares, the gameboy save file that will be loaded is <gb_rom_filename_without_extension>.ram)

This pull request replaces https://github.com/DragonMinded/libdragon/pull/579

risingPhil committed:
Implement --controller<x> options in ed64romconfig to change the respective Advanced Homebrew Header fields

Modifying these fields may influence emulator behaviour, such as Ares'

The context in which I implemented this is to make debugging N64 Transfer Pak code easier with Ares Emulator:
After using the --controller1 n64,pak=transfer option on your rom, Ares will ask for a gameboy rom for the emulated transfer pak
everytime you load the rom. (Note: In Ares, the gameboy save file that will be loaded is <gb_rom_filename_without_extension>.ram)

risingPhil committed:
PR feedback: make --controller<x> short form options POSIX compliant

I got some feedback to make the --controller<x> short form options POSIX compliant by reducing them to a single character.

Therefore I changed those short form options to -1, -2, -3 and -4.

meeq commented on tools/ed64romconfig.c:
```diff
@@ -51,9 +67,23 @@ int print_usage(const char * prog_name)
+	fprintf(stderr, "\t    n64,pak=transfer            N64 controller with Transfer Pak\n");
+	fprintf(stderr, "\t    nothing                     Nothing attached to this port\n");
+	fprintf(stderr, "\t    mouse                       N64 mouse\n");
+	fprintf(stderr, "\t    vru                         VRU\n");
```
Tangential: Does Ares make a distinction between VRU/VRS?

risingPhil replied:
No. As a matter of fact it appears to be currently unimplemented.

<ares-repo>/mia/medium/nintendo-64.cpp (line 740) says "not supported yet" for all options starting at the N64 mouse. In practice, that means only the "pak" and default options are actually supported by Ares for the moment.

Btw, I was aware of this going in. But not being supported right now doesn't mean they never will. I just added all the valid options according to https://n64brew.dev/wiki/ROM_Header#Advanced_Homebrew_ROM_Header

meeq replied:
I assume an emulator could make the distinction by the ROM's region. That would prevent Japanese games from being able to use VRU, and English games from using VRS, but that's a pretty extreme edge case that I don't see being a practical limitation.

meeq commented on tools/ed64romconfig.c:
```diff
@@ -51,9 +67,23 @@ int print_usage(const char * prog_name)
+	fprintf(stderr, "\t    n64,pak=rumble              N64 controller with Rumble Pak\n");
+	fprintf(stderr, "\t    n64,pak=controller          N64 controller with Controller Pak\n");
+	fprintf(stderr, "\t    n64,pak=transfer            N64 controller with Transfer Pak\n");
+	fprintf(stderr, "\t    nothing                     Nothing attached to this port\n");
```
I can think of a couple other aliases that I would try before "nothing" if I wasn't sure of the syntax:
* none
* empty
* disconnected

Consistency with save_type would suggest that "none" is the preferable option here.

risingPhil replied:
Okay. Makes sense. I just followed the name from the n64brew page.

But I'll change it.

risingPhil replied:
done

meeq commented on tools/ed64romconfig.c:
```diff
@@ -51,9 +67,23 @@ int print_usage(const char * prog_name)
+	fprintf(stderr, "\t    nothing                     Nothing attached to this port\n");
+	fprintf(stderr, "\t    mouse                       N64 mouse\n");
+	fprintf(stderr, "\t    vru                         VRU\n");
+	fprintf(stderr, "\t    ngc                         Gamecube controller\n");
```
`ngc` is less clear than `gamecube`

risingPhil replied:
Okay, I'll change it.

thekovic replied:
I'll just add that official styling is "GameCube" (capital C) so the printed text should follow that.

risingPhil replied:
done both things now ;-)

meeq commented on tools/ed64romconfig.c:
```diff
@@ -51,9 +67,23 @@ int print_usage(const char * prog_name)
+	fprintf(stderr, "\t-r, --regionfree                Declare region-free ROM.\n");
+	fprintf(stderr, "\t-1, --controller1 <type>        Define controller 1 hardware type. <type> should be one of:\n");
+	fprintf(stderr, "\t    default                     Emulators should follow standard config for this port.\n");
+	fprintf(stderr, "\t    n64,pak=rumble              N64 controller with Rumble Pak\n");
```
Should probably have a `n64` (no `,pak=`) option for completeness

risingPhil replied:
Hmm... Based on https://n64brew.dev/wiki/ROM_Header#Advanced_Homebrew_ROM_Header, It seems that that might be the "default" option.

...which brings us back to your previous comment. So up to you: should I rename the "default" argument to "n64"? Or just omit it entirely?

meeq replied:
I'd say rename "default" to "n64".

meeq commented on tools/ed64romconfig.c:
```diff
@@ -51,9 +67,23 @@ int print_usage(const char * prog_name)
+	fprintf(stderr, "\t-c, --rtc                       Declare real-time clock support.\n");
+	fprintf(stderr, "\t-r, --regionfree                Declare region-free ROM.\n");
+	fprintf(stderr, "\t-1, --controller1 <type>        Define controller 1 hardware type. <type> should be one of:\n");
+	fprintf(stderr, "\t    default                     Emulators should follow standard config for this port.\n");
```
Why would you ever explicitly set default instead of just omitting the parameter?

risingPhil replied:
I suppose that's true. I'll remove it. ;-)

risingPhil replied:
closing because of comment below

meeq commented on n64.mk:
```diff
@@ -8,6 +8,7 @@ N64_ROM_RTC = # Set to true to enable the Joybus Real-Time Clock
 N64_ROM_REGIONFREE = # Set to true to allow booting on any console region
 N64_ROM_REGION = # Set to a region code (emulators will boot on a specific console region)
 N64_ROM_ELFCOMPRESS = 1 # Set compression level of ELF file in ROM
+N64_ROM_CONTROLLER_TYPE1 = # Sets controller type in the Advanced Homebrew Header. This influences emulator behaviour such as Ares'
```
Suggestion: rename to `N64_ROM_CONTROLLER1`

Also, consider adding 2, 3, 4

risingPhil replied:
Okay :-)

risingPhil replied:
done

meeq commented on tools/ed64romconfig.c:
```diff
@@ -170,6 +309,14 @@ int main(int argc, char *argv[])
 	fseek(write_file, VERSION_OFFSET, SEEK_SET);
 	fwrite(&config, 1, VERSION_SIZE, write_file);
+	fseek(write_file, CONTROLLERTYPE1_OFFSET, SEEK_SET);
+	fwrite(&controller_type1, 1, 1, write_file);
```
["When in Rome, do as the Romans do"](https://tommcfarlin.com/coding-standards-and-guidelines/)

A `CONTROLLERTYPE_SIZE` constant is not strictly necessary, but I do think it is "better".

risingPhil replied:
Okay, I'll change it.

risingPhil replied:
done

risingPhil committed:
PR feedback from @meeq on the --controller<x> options pull request (#580)

n64.mk:
- rename N64_ROM_CONTROLLER_TYPE1 to N64_ROM_CONTROLLER1
- add the same for controller 2, 3 and 4

ed64romconfig.c:
- rename the "--controller default" option to "--controller n64"
- replace the "nothing" controller type with "none"
- rename the "ngc" controller type to "gamecube"
- add CONTROLLERTYPE_SIZE constant and use it for the fwrite() calls.

risingPhil committed:
Fix another PR comment - Correct capitalization of GameCube to official form.

meeq commented on n64.mk:
```diff
@@ -62,6 +66,7 @@ N64_TOOLFLAGS += $(if $(N64_ROM_REGION),--region $(N64_ROM_REGION))
 N64_ED64ROMCONFIGFLAGS =  $(if $(N64_ROM_SAVETYPE),--savetype $(N64_ROM_SAVETYPE))
 N64_ED64ROMCONFIGFLAGS += $(if $(N64_ROM_RTC),--rtc) 
 N64_ED64ROMCONFIGFLAGS += $(if $(N64_ROM_REGIONFREE),--regionfree)
+N64_ED64ROMCONFIGFLAGS += $(if $(N64_ROM_CONTROLLER_TYPE1),--controller1 $(N64_ROM_CONTROLLER_TYPE1))
```
Missing 2, 3, 4

risingPhil replied:
shoot. Missed those.

I've added them now. Good catch!

risingPhil committed:
PR feedback: add missing N64_ED64ROMCONFIGFLAGS for controller type 2,3,4

n64.mk:

Add missing N64_ED64ROMCONFIGFLAGS for controller types 2,3 and 4.
